data/data/baremetal/masters/main.tf

```diff
@@ -0,0 +1,53 @@
+# This null resource is used to wait for the Ironic API to become available before trying to create
+# the master hosts.
+resource "null_resource" "api-wait" {
+  provisioner "local-exec" {
```

Comment:
FYI, this will not be allowed as this runs commands on user's host.

Reply:
We weren't particularly happy with this, but had trouble finding another path. Do you have a suggestion for a way we can do this? Once Ironic is on the bootstrap node, we have the problem that one provider's entire infrastructure is brought up by another one.  This isn't really well supported by terraform https://github.com/hashicorp/terraform/issues/2430, and mostly people use hacks like this.

Could we do remote-exec on the bootstrap?

Reply:
My 2c: The installer should also have a container image as part of the payload that is executed on the bootstrap.  That would pave the way to fixing a *lot* of things, including e.g. having masters provisioned by machine API, etc.

Reply:
>  That would pave the way to fixing a lot of things, including e.g. having masters provisioned by machine API, etc.

how does that tie into your comment?

Reply:
edit: Removed that bit since it was distracting

Reply:
> My 2c: The installer should also have a container image as part of the payload that is executed on the bootstrap. Basically it should be as ergonomic to commit code into the installer git to run on the bootstrap as it is to run on the "deploying machine".

Can you clarify what this would do for us? I'm not entirely sure I understand.  The only thing being run on the deploying machine is curl to check if Ironic is available.

We are in the process of moving Ironic as a container onto the bootstrap VM.  This local exec exists for that reason: this resource waits for the Ironic API to be available, and the master resources have a depends_on.  

This prevents terraform from trying to create the masters until Ironic is available on the bootstrap.  This is a hack many people use when one provider is responsible for bringing up the infrastructure used by another.

Reply:
The PR for the Ironic on bootstrap is https://github.com/openshift-metal3/kni-installer/pull/100, it's still being worked on by @hardys.

> That would pave the way to fixing a lot of things, including e.g. having masters provisioned by machine API, etc.

If masters were provisioned by machine API, then this would avoid a lot of headaches since we wouldn't need terraform for Ironic at all.

Reply:
Also note that openshift-metal3/kni-installer#100 doesn't remove the need to wait for ironic to be ready - just starting the container on the boostrap VM isn't enough, we need the API to be up (which doesn't take long), and also for the RHCOS and deploy ramdisk/kernel images to be downloaded (which can take a while).

In my PR I'm using the same curl approach to wait for all-the-things in terraform, so if that's unacceptable we'll need to agree a solution that is.

Reply:
Another alternative would be to refactor the pkg/asset/cluster code to support the idea of multiple terraform stages if a platform requires it, i.e. 2 distinct `terraform apply`'s where one brings up bootstrap (with Ironic), and the second utilizing Ironic to create the masters.

Reply:
Just noting that we've removed this again in openshift-metal3/kni-installer#133 so it'll be gone when this PR is refreshed